cli/flox/src/commands/mod.rs

```diff
@@ -1627,29 +1627,36 @@ async fn maybe_migrate_environment_to_v1(
 /// If confirm_upgrade_future is None, the user can't be prompted to confirm an upgrade.
 async fn maybe_migrate_environment_to_v1_inner(
     flox: &Flox,
-    environment: &mut ConcreteEnvironment,
+    concrete_environment: &mut ConcreteEnvironment,
     confirm_upgrade_future: Option<impl Future<Output = Result<bool>>>,
 ) -> Result<(), MigrationError> {
-    if let ConcreteEnvironment::Managed(ref environment) = environment {
-        if environment
-            .has_local_changes(flox)
-            .map_err(EnvironmentError::ManagedEnvironment)?
-        {
-            Err(EnvironmentError::ManagedEnvironment(
-                ManagedEnvironmentError::CheckoutOutOfSync,
-            ))?;
-        }
-    }
-
-    let description = environment_description(environment)?;
-    let environment = environment.dyn_environment_ref_mut();
+    let description = environment_description(concrete_environment)?;
 
     if flox.catalog_client.is_none() {
         return Ok(());
     }
     // The user answered the prompt to upgrade the environment
     let mut confirmed_upgrade = false;
-    if let Some(migration_info) = environment.needs_migration_to_v1(flox)? {
+    if let Some(migration_info) = concrete_environment
+        .dyn_environment_ref_mut()
+        .needs_migration_to_v1(flox)?
+    {
+        // If the migration requires changes to a manifest, check if the environment is in sync
+        // before asking whether to migrate
+        // (which would subsequently fail due to changes in the manifest).
+        if migration_info.needs_manifest_migration {
+            if let ConcreteEnvironment::Managed(ref environment) = concrete_environment {
+                if environment
+                    .has_local_changes(flox)
+                    .map_err(EnvironmentError::ManagedEnvironment)?
+                {
+                    Err(EnvironmentError::ManagedEnvironment(
+                        ManagedEnvironmentError::CheckoutOutOfSync,
+                    ))?;
+                }
+            }
+        }
+
```

Comment:
blocking: I think we need a unit test for the case we had bad behavior for?

Reply:
you're right, added a test